kong/vault_service.go

```diff
@@ -0,0 +1,158 @@
+	// Delete deletes a Vault in Kong
+	Delete(ctx context.Context, nameOrID *string) error
+	// List fetches a list of Vaults in Kong.
+	List(ctx context.Context, opt *ListOpt) ([]*Vault, *ListOpt, error)
+	// ListAll fetches all Vaults in Kong.
+	ListAll(ctx context.Context) ([]*Vault, error)
+}
+
+// VaultService handles Vaults in Kong.
+type VaultService service
+
+// Create creates a Vault in Kong
+// If an ID is specified, it will be used to
+// create a Vault in Kong, otherwise an ID
+// is auto-generated.
+func (s *VaultService) Create(ctx context.Context, vault *Vault) (*Vault, error) {
+	if vault == nil {
+		return nil, fmt.Errorf("cannot create a nil vault")
+	}
+
+	endpoint := "/vaults"
+	method := "POST"
+	if vault.ID != nil {
+		endpoint = endpoint + "/" + *vault.ID
+		method = "PUT"
+	}
+	req, err := s.client.NewRequest(method, endpoint, nil, vault)
+	if err != nil {
+		return nil, err
+	}
+
+	var createdVault Vault
+	_, err = s.client.Do(ctx, req, &createdVault)
+	if err != nil {
+		return nil, err
+	}
+	return &createdVault, nil
+}
+
+// Get fetches a Vault in Kong.
+func (s *VaultService) Get(ctx context.Context, prefixOrID *string) (*Vault, error) {
+	if isEmptyString(prefixOrID) {
+		return nil, fmt.Errorf("prefixOrID cannot be nil for Get operation")
+	}
+
+	endpoint := fmt.Sprintf("/vaults/%v", *prefixOrID)
+	req, err := s.client.NewRequest("GET", endpoint, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	var vault Vault
+	_, err = s.client.Do(ctx, req, &vault)
+	if err != nil {
+		return nil, err
+	}
+	return &vault, nil
+}
+
+// Update updates a Vault in Kong
+func (s *VaultService) Update(ctx context.Context, vault *Vault) (*Vault, error) {
+	if vault == nil {
+		return nil, fmt.Errorf("cannot update a nil vault")
+	}
+
+	if isEmptyString(vault.ID) {
+		return nil, fmt.Errorf("ID cannot be nil for Update operation")
+	}
+
+	endpoint := fmt.Sprintf("/vaults/%v", *vault.ID)
+	req, err := s.client.NewRequest("PATCH", endpoint, nil, vault)
+	if err != nil {
+		return nil, err
+	}
+
+	var updatedVault Vault
+	_, err = s.client.Do(ctx, req, &updatedVault)
+	if err != nil {
+		return nil, err
+	}
+	return &updatedVault, nil
+}
+
+// Delete deletes a Vault in Kong
+func (s *VaultService) Delete(ctx context.Context, prefixOrID *string) error {
+	if isEmptyString(prefixOrID) {
+		return fmt.Errorf("prefixOrID cannot be nil for Delete operation")
+	}
+
+	endpoint := fmt.Sprintf("/vaults/%v", *prefixOrID)
+	req, err := s.client.NewRequest("DELETE", endpoint, nil, nil)
+	if err != nil {
+		return err
+	}
+
+	_, err = s.client.Do(ctx, req, nil)
+	return err
+}
+
+// List fetches a list of Vaults in Kong.
+// opt can be used to control pagination.
+func (s *VaultService) List(ctx context.Context,
+	opt *ListOpt,
+) ([]*Vault, *ListOpt, error) {
+	data, next, err := s.client.list(ctx, "/vaults", opt)
+	if err != nil {
+		return nil, nil, err
+	}
+	var vaults []*Vault
+	for _, object := range data {
+		b, err := object.MarshalJSON()
```

Comment:
It seems that the underlying type of [`json.RawMessage`](https://pkg.go.dev/encoding/json#RawMessage) is `[]byte` which means that it should be possible to "just use it" as `json.Unmarshal()` parameter. This would save us the unnecessary marshalling.

Can't test it now because it seems I need to do some shenanigans to start kong locally (manually) to do so.

Reply:
Good catch!